config: handle CRLF-only lines and BOM

pks-t opened the pull request:
Fixes #4517

csware commented:
This fixes the issue.

pks-t commented:
Amended the PR to also fix handling of the BOM.

csware commented:
Works. Thanks!

ethomson commented on src/config_parse.c:
```diff
@@ -217,7 +217,7 @@ static int skip_bom(git_parse_ctx *parser)
 	git_buf buf = GIT_BUF_INIT_CONST(parser->content, parser->content_len);
 	git_bom_t bom;
-	int bom_offset = git_buf_text_detect_bom(&bom, &buf, parser->content_len);
+	int bom_offset = git_buf_text_detect_bom(&bom, &buf, 0);
```
I understand your confusion.  It appears that we never use `git_buf_text_detect_bom` to detect a BOM with a non-0 offset.

It's not clear to me why this was ever added in the first place - when would we try to detect a BOM in the midst of a buffer?

Whatever the reason, it sounds like we've refactored things to always look at the beginning of the buffer.  Should we drop this confusing third arg?

pks-t replied:
I think so. I even wonder if there is any usecase with a mid-buffer-BOM, considering they are only allowed at the beginning of a file.

tiennou commented on src/config_parse.c:
```diff
@@ -475,6 +475,10 @@ int git_config_parse(
 
+		/*
+		 * Get either first non-whitespace character or, if that does
+		 * not exist, the first whitespace character
```
`because we need to preserve whitespace when writing back` ?

Maybe `We skip over whitespace until we find something meaningful, otherwise there's only whitespace and we need to preserve it when writing back`.

pks-t replied:
True, that would certainly go a long way in explaining what happens here. Thanks, will amend!

pks-t commented:
Amended. Added @tiennou's proposed improvement for the comment and removed the `offset` parameter of the BOM function as proposed by @ethomson. Thanks to both of you!

ethomson commented:
Thanks @csware for the report, @tiennou and @pks-t for the collaboration to get it fixed.  👍
